lib/models/BucketPolicy.js

```diff
@@ -0,0 +1,127 @@
+ *              "Effect": "Allow",
+ *              "Principal": "*",
+ *              "Action": "s3:*",
+ *              "Resource": "arn:aws:s3:::examplebucket/bucket2/object"
+ *          },
+ *          {
+ *              "Sid": "Statement id",
+ *              "Effect": "Deny",
+ *              "Principal": {
+ *                  "AWS": ["arn:aws:iam::<account_id>", "different_account_id"]
+ *              },
+ *              "Action": [ "s3:*" ],
+ *              "Resource": [
+ *                  "arn:aws:s3:::examplebucket", "arn:aws:s3:::otherbucket/*"],
+ *              "Condition": {
+ *                 "StringNotLike": {
+ *                      "aws:Referer": [
+ *                          "http://www.example.com/", "http://example.com/*"]
+ *                  }
+ *              }
+ *          }
+ *      ]
+ *  }
+ */
+
+class BucketPolicy {
+    /**
+     * Create a Bucket Policy instance
+     * @param {string} json - the json policy
+     * @return {object} - BucketPolicy instance
+     */
+    constructor(json) {
+        this._json = json;
+        this._policy = {};
+    }
+
+    /**
+     * Get the bucket policy
+     * @return {object} - the bucket policy or error
+     */
+    getBucketPolicy() {
+        const policy = this._getPolicyJSON();
+        return policy;
+    }
+
+    /**
+     * Get the bucket policy array
+     * @return {object} - contains error if policy validation fails
+     */
+    _getPolicyJSON() {
+        const policy = {};
+        if (!this._json || this._json === '') {
+            policy.error = errors.MalformedPolicy.customizeDescription(
+                'request json is empty or undefined');
+            return policy;
+        }
+        const validSchema = validateResourcePolicy(this._json);
+        if (validSchema.error) {
+            policy.error = validSchema.error;
+            return policy;
+        }
+        this._getStatementArray();
+        const valAcRes = this._validateActionResource();
+        if (valAcRes.error) {
+            policy.error = valAcRes.error;
+            return policy;
+        }
+
+        return this._policy;
+    }
+
+    _getStatementArray() {
+        this._policy = JSON.parse(this._json);
+        if (!Array.isArray(this._policy.Statement)) {
+            const statement = this._policy.Statement;
+            this._policy.Statement = [statement];
+        }
+    }
+
+    /**
+     * Validate action and resource are compatible
+     * @return {error} - contains error or empty obj
+     */
+    _validateActionResource() {
+        const obj = {};
+        this._policy.Statement.forEach(s => {
+            const actions = typeof s.Action === 'string' ?
+                [s.Action] : s.Action;
+            const resources = typeof s.Resource === 'string' ?
+                [s.Resource] : s.Resource;
+            const objectAction = actions.some(a => a.includes('Object'));
+            const allAction = actions.some(a => a.includes('*'));
```

Comment:
I haven't tried it out, but it sounds valid also use wildcards as part of the action name e.g. s3:Get* which would then only apply to a subset of actions. I'm wondering because in the AWS docs it shows "Action": "iam:*AccessKey*" as valid.

Reply:
Oof, you're right about the `s3:Get*` thing, that's a great catch. I didn't realize that was valid. But for the `iam:` action, do you have a link to the example? As far as I can tell, only S3 operations are valid as actions. Even if the `iam` prefix is valid tho, we're still only going to support S3 actions for right now

Reply:
Oh yes, I agree it'd need to start with "s3". It's just that I came across that example [here](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_action.html) using "iam" so figured it could apply similarly.

Reply:
aha I see. I think that example page is for IAM/user policies specifically :)

Reply:
Ok, I've accounted for an asterisk in the action and added some new tests.